motionplan/inverseKinematics.go

```diff
@@ -2,23 +2,232 @@ package motionplan
 
 import (
 	"context"
+	"math"
+	"runtime"
+	"sort"
+
+	"github.com/edaniels/golog"
+	"go.viam.com/utils"
 
 	"go.viam.com/rdk/referenceframe"
-	spatial "go.viam.com/rdk/spatialmath"
+	"go.viam.com/rdk/spatialmath"
+)
+
+// default values for inverse kinematics.
+const (
+	// If an IK solution scores below this much, return it immediately.
+	defaultMinIkScore = 0.
+
+	// Number of IK solutions that should be generated before stopping.
+	defaultSolutionsToSeed = 50
 )
 
-// InverseKinematics defines an interface which, provided with a goal position and seed inputs, will output all found
-// solutions to the provided channel until cancelled or otherwise completes.
-type InverseKinematics interface {
-	// Solve receives a context, the goal arm position, and current joint angles.
-	Solve(context.Context, chan<- []referenceframe.Input, spatial.Pose, []referenceframe.Input, Metric, int) error
+var defaultNumThreads = runtime.NumCPU() / 2
+
+type ikOptions struct {
+	constraintHandler
+	extra map[string]interface{}
+
+	// Metric by which to measure nearness to the goal
+	metric Metric
+
+	// Solutions that score below this amount are considered "good enough" and returned immediately
+	MinScore float64 `json:"min_ik_score"`
+
+	// Max number of ik solutions to consider
+	MaxSolutions int `json:"max_ik_solutions"`
+
+	// Number of cpu cores to use
+	NumThreads int `json:"num_threads"`
```

Comment:
This should not be in `ikOptions`, but rather in a more generic struct that `ikOptions` inherits from, along with `extra` and `constraintHandler`. Otherwise, it will be impossible for planners to use these things without creating an `ikOptions`, regardless of whether the planner does IK.

Reply:
In the same way you don't want to prematurely optimize for non-RRT planners, I would rather not make this change and cater to imagined future planners that don't use IK until we actually have an instance of this.

Reply:
Does this even need to have `ik` in the name? What are other planners going to produce that aren't already captured in metrics, scores, and solutions?

Reply:
It's fine not to cater to nonexistent planners, but we should organize the fields such that they make sense where they are used; it doesn't make sense to have a motion planner look in its IK options for the number of CPU threads to use.

Reply:
`MinScore` and `MaxSolutions` are the only IK specific things, so if we want IK specific options, those are what should be in it.

The other fields should migrate back to base `plannerOptions`, which IKoptions should inherit from, not vice versa.